Proposal for automatic discovery of content routers

willscott opened the pull request:
This follows the previously circulated proposal outline at https://hackmd.io/bh4-SCWfTBG2vfClG0NUFg

A basic motivation is included in the PR - but essentially this is the best path I've heard for reducing our dependence on hydras as a centrally operated choke point for moving the bulk of the IPFS network beyond sole reliance on the current KAD DHT.

mxinden commented on IPIP/0000-content-router-discovery.md:
```diff
@@ -0,0 +1,250 @@
+* The server will have a parameter for a number of servers it wants to return
+to content routing queries. By default this will be 10. (This default is picked
+as the result of modeling router propagation). It will iterate through it's
+list of known content routers, hashing htem against the bloom filter and
```
```suggestion
list of known content routers, hashing them against the bloom filter and
```

ajnavarro commented on IPIP/0342-content-router-discovery.md:
```diff
@@ -0,0 +1,246 @@
+
+## Test fixtures
+
+TK is a CID currently only available through the content routing system,
```
TK? do you mind to uncompress the alias? :)

ajnavarro commented on IPIP/0342-content-router-discovery.md:
```diff
@@ -0,0 +1,246 @@
+
+This design is self-contained - it does not require standing up additional
+infrastructure or making additional connections for discovery but rather
+gossips routers over existing peer connections.
```
this is the first time gossip is mentioned. Should we be more specific in the `Detailed design` section about the protocol and how nodes will be interconnected?

ajnavarro commented on IPIP/0342-content-router-discovery.md:
```diff
@@ -0,0 +1,246 @@
+
+Nodes which do not upgrade to support this IPIP will be limited to the sub-set of
+content available in the DHT. this will potentially degrade over time as more
+large providers limit their publishing per the IPNI ingestion protocol.
```
IPNI?

willscott replied:
https://github.com/ipni

ajnavarro commented on IPIP/0342-content-router-discovery.md:
```diff
@@ -0,0 +1,246 @@
+not be directly discovered. Instead, the gossip discovery protocol is
+ambiently discovered in much the same way as circuit relays.
+
+#### Advertisement in the DHT
```
The good things about advertising using the DHT are:
- Network that is already there, no need to create a new protocol to "provide" new providers instead of CIDs.
- You can provide associating your provider with a specific root CID content. I seriously doubt that all providers will be eager to provide all CIDS in the universe.

ajnavarro reviewed (commented):
I'm missing a way to link content with the provider because I seriously doubt that all parties will be eager to provide all the CIDs in the universe, they will focus on providing the content they care about.

We could link root CID/CIDs with the provider to let know to nodes where they can find the DAG for specific content.

willscott commented:
> I'm missing a way to link content with the provider because I seriously doubt that all parties will be eager to provide all the CIDs in the universe, they will focus on providing the content they care about.
> 
> We could link root CID/CIDs with the provider to let know to nodes where they can find the DAG for specific content.

This is a philosophical disagreement about what a 'content router' is.
I'm going to refer you to the [broad guidance](https://www.youtube.com/watch?v=t0u-xMwWra4) that was presented over the summer for the evolution of content routing.

We currently have a couple examples of content routers that do have all the CIDs in the universe, and we do not have convincing examples of, or a definition for, sub-content-routers as you're proposing here.

Why are we trying to compromise to a much-harder-to-make-work complexity without trying for the thing that makes sense and the direction we're heading first?

willscott commented on IPIP/0342-content-router-discovery.md:
```diff
@@ -0,0 +1,246 @@
+### 1. content-routing as a libp2p protocol
+
+IPFS nodes will advertise and coordinate discover of content routers using a
+new libp2p protocol advertised as "".
```
do you have thoughts on what a good name would be?
maybe '/ipfs/content-router-discovery/1.0.0' ?

willscott commented on IPIP/0342-content-router-discovery.md:
```diff
@@ -0,0 +1,246 @@
+* reliability - how many good vs bad responses has this router responded
+with. This statistic should be windowed, such that the client can calculate
+it in terms of the last week or month.
+* performance - how quickly does this router respond.
```
Yes, this would be windowed. I was imagining a window of ~ "last week" by default, but this seems like a good candidate to evaluate through simulation.

willscott commented on IPIP/0342-content-router-discovery.md:
```diff
@@ -0,0 +1,246 @@
+list of known content routers, hashing them against the bloom filter and
+selecting the top routers that are not already known to the client. It will
+return this list, along with it's reliability score for each. This response
+is structured as an IPLD list lists, conceptually:
```
a json or cbor array are both examples that would fulfill this, I'll leave it more generic but i have a hard time imagining we'd encode this in a way that wouldn't conform to being considered an IPLD list

ajnavarro commented:
@willscott I think more than a philosophical disagreement is a physical one.

Right now we are able to keep all the CIDs on the network in one provider for two reasons:
- The network is relatively small
- Right now we have only two places where nodes are providing CIDs: DHT and Bitswap, so it is easier to get that information because is only in two specific places.

When we start to have different ways of providing CIDs, will be near impossible to have everything replicated by everyone. 

Also when the network scales, having centralized all the information in several places will be quite challenging and costly. But on the other hand, allowing both approaches (providing everything vs providing a subset of the CIDs) will have for sure their use cases for people with not a huge amount of money to maintain big providers.

willscott commented:
We currently have providers stepping up to provide full replicas of a content routing database. That is what network indexers have been doing over the last year.
I don't see the physical issue here: indexers are handling trillions of records, vs the 100 billion in the DHT, and the scale we expect as we grow will still fit at the level of a single rack in a data center.

In designing delegated routing so far, the eye has been towards a design where delegated routers need to fall back and do the additional work of querying other routers in order to collect a full replica if they don't possess it themselves, rather than making that the end kubo node's responsibility, as that leads to an untenable performance and decision process for end user nodes that are not equipped to handle that.

I'm not entirely sure of your counter proposal here: I think there are very strong counter arguments against both https://github.com/ipfs/specs/pull/322 - which compromises trying to be a content addressed network, or limited DHT providing (e.g. to roots) - which still couldn't handle the current indexer database scale.

aschmahmann commented on IPIP/0342-content-router-discovery.md:
```diff
@@ -0,0 +1,267 @@
+support this prioritization without leaking the exact list of known content
+routers that the client already knows.
+
+* The size of the bloom filter is chosen by the client. It is sized such
```
Unfortunately GitHub doesn't allow threads not tied to a line, but wanted to add some thoughts to this discussion https://github.com/ipfs/specs/pull/342#issuecomment-1316854045 in a way that responses would be easy to trace.

1) Per @ajnavarro's comment https://github.com/ipfs/specs/pull/342#issuecomment-1318830270 I too find it a little hard to believe that there will be so many routers each providing full replicas of all data tracked by IPNI that a bloom filter would be required given that running these servers is expensive and incentivization is IIUC mostly TBD (I think @guseggert had some napkin math here showing the large costs around storing 10^15 CIDs even if we exclude bandwidth costs). That being said this is `#not-this-ipips-problem`. If the IPNI team thinks thousands of nodes all over the world will spring up hosting PBs of data and that lack of consistency between replicas isn't going to cause problems with the evaluation criteria that clients use that problem seems to live elsewhere.
2) Whether or not #322 is a good/bad idea is also `#not-this-ipips-problem` since this describes how to find routers for a given content routing system (i.e. IPNI) not whether it should be passable as a hint.
    - As an aside my 2c is that you've got to be careful here to not break IPLD properties if you go this route as I've flagged in #322, however it's potentially useful to add hints as long as they're not mandatory.

willscott commented on IPIP/0342-content-router-discovery.md:
```diff
@@ -0,0 +1,267 @@
+
+### 1. content-routing as a libp2p protocol
+
+IPFS nodes will advertise and coordinate discover of content routers using a
+new libp2p protocol advertised as "/ipfs/content-router-discovery/1.0.0".
```
> This currently seems to be specifically for IPNI routers so at best this is /ipfs/ipni-discovery/1.0.0

this is for discovery of content routers per the delegated content router API - what about this is IPNI specific?

aschmahmann replied:
I'm confused as to how the current draft addresses the issues here:

> will have knowledge of the entire CID space

This line (from the spec) seems problematic even in the IPNI case. IPNI != the entire CID space which starts to make implementations complicated. The idea that one routing system is going to cover every use case is IMO not the way to go (and also the reason why there's even discussion of a delegated content routing API rather than just an IPNI API).

---

An example issue is to say we have 4 routers:
1. cid.contact (IPNI)
2. FilSwan (IPNI)
3. routing.delegate.ipfs.io (proxies DHT + IPNI)
4. ipfsdht.delegate.ipfs.io (proxies DHT, this is the only one that doesn't exist today, but certainly could)

While router 3 provides strictly more information than routers 1 or 2 it's also likely to be slower than them. It seems optimal to either contact 3 or contact 1/2 + 4 in parallel. For a node running its own DHT client would contact routers 1 or 2 and never contact 4. However, a naïve implementation may just result in all requests going through router 3 as it has the most CIDs covered which is not good. Perhaps a classification algorithm would be able to tease-out the optimizations here without further protocol adjustments, but that seems like a lot of complexity that could be alleviated by a small protocol adjustment.

This case also seems more problematic than the one that's been resolved by flagging router "type" like `content-routing`, `peer-routing`, etc. since support for a given API (content/peer routing) can be discovered with a single query to the endpoint whereas this requires a bunch of code complexity.

This seems like it'd be largely resolvable based on allowing users to query and return a set of named routing systems, or just calling this the "IPNI discovery system" so that routers like 3 + 4 know not to participate. I'd rather the former, but understand the latter.

While I wouldn't be surprised if down the road we also ended up requiring some of that  ML-style classification code anyway I suspect walking down that path now is premature and likely to cause us problems.

willscott replied:
* It seems like there's a lot of complexity in expressing this 'composition complexity' either directly or through classification. we don't have this problem today so I would prefer to defer this sort of grammar to a subsequent IPIP - you worry that 3 would do better than the others, but I would argue that would be incentive for the IPNI team to build what i think you have previously called 'radar' to incorporate DHT results into IPNI such that 1,2,3 are all equal :)

* As you say, "IPNI != the entire CID space". I think it's a mistake to limiting our framing of this to an "IPNI discovery system" when it is simply discovering 'the most complete' available content routers. We're trying to be inclusive/general here - and I don't see huge harm in calling this content routers.

aschmahmann replied:
> we don't have this problem today

I guess it depends what you mean by "today". I would like libraries like https://github.com/libp2p/js-libp2p-delegated-content-routing to switch to the latest content routing API (or have alternative which have switched) at which point js-libp2p in browsers should be able to leverage both the DHT and IPNI to get data from any peer that speaks wss/webtransport/webrtc.

Ideally they could use this protocol for discovery rather than hardcoding a DHT resolution endpoint.

@lidel could probably speak more about desired timings here.

> but I would argue that would be incentive for the IPNI team to build what i think you have previously called 'radar' to incorporate DHT results into IPNI such that 1,2,3 are all equal :)

That's cool and would certainly resolve at least this use case 😄.

---

> We're trying to be inclusive/general here
❤️

>  I think it's a mistake to limiting our framing of this to an "IPNI discovery system" when it is simply discovering 'the most complete' available content routers.

That's an interesting framing. By pushing for the "most complete set" it seems like you're essentially trying to get routers to compete for attention and content and make it so that only a single request to a single system needs to happen for clients to get what they need. If this is how the system evolves this is very nice to client machines.

However, if routers try and cut costs or code complexity by serving more-specific data (e.g. only data advertised over a specific pubsub channel, only data put in the IPFS Public DHT, only data from the BitTorrent network, ...) then the client code could start becoming problematic as it tries to figure out who to ask, without spamming all the routers.

I'm more cautious in advocating for the latter, but could see this go either way 😅. As long as the more immediate case around DHT + IPNI data being available to browser nodes is covered I'm happy 😄.

lidel replied:
+1 that delegated  DHT  is the core use case for IPFS  on Web Platform (JS in HTML in web browser)

- The gist of https://github.com/libp2p/js-libp2p-delegated-content-routing story is that is uses `/api/v0/dht` from Kubo RPC and we want to move away from that model. 
- Switching to HTTP API at `routing.delegate.ipfs.io` (proxies DHT + IPNI) is an easy win, and we would want to do this ASAP.
- Having ambient discovery via bootstrap nodes talking `/wss` and `/webtransport` will allow for basic resiliency / redundancy

lidel commented on IPIP/0342-content-router-discovery.md:
```diff
@@ -0,0 +1,267 @@
+content available in the DHT. this will potentially degrade over time as more
+large providers limit their publishing per the IPNI ingestion protocol.
+
+Nodes may limit their complexity through a hard-coded list of known content
+routers, essentially limiting their implementation to design section 3 of this
+IPIP. In doing so, they may limit their risk of exposure to malicious parties.
+They risk being out of date and to offer sub-optimal performance through their
+failure to discover additional near-by content routing instances.
```
>  hard-coded list [..] In doing so, they may limit their risk of exposure to malicious parties.

Hardcoding HTTP URLs is not always "limiting risk", it is changing the threat model by introducing a single point of failure as an obvious denial of service vector.  Depending on use case, may actually **increase the risks to the user**. 
Better to remove this statement.

```suggestion
Nodes may limit their complexity through a hard-coded list of known content
routers, essentially limiting their implementation to design section 3 of this
IPIP.  This comes at a price: (1) hard-coded routers become easy targets 
for denial of service attacks, decreasing the resilliency of the entire setup;
(2) nodes risk being out of date and to offer sub-optimal performance through their
failure to discover additional near-by content routing instances.
```

lidel commented on IPIP/0342-content-router-discovery.md:
```diff
@@ -0,0 +1,267 @@
+IPFS nodes, and the other side are the bootstrap and core-infrastructural
+nodes with high connectivity in the network.
+
+### 1. content-routing as a libp2p protocol
```
I suspect the idea is for `/dnsaddr/bootstrap.libp2p.io` (or any other bootstrapper set by JS user, as long its  `/webtransport` or `/wss`)  to speak this new protocol, avoiding hardcoding anything new.

willscott replied:
what prevents them from participating in this protocol as described?
browser nodes will need to contact to other existing nodes, as they do today. they would learn about the existence of content routers through those same channels via the new protocol, and could then make use of them.

lidel replied:
> what prevents them from participating in this protocol as described?

CORS. If the only type of router this protocol returns is HTTP URL, then by default JS-IPFS running on a website won't be able to read data via cross-origin requests to the discovered router due to [CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) limitations. 

We have two ways of solving the problem:
1.  (easy spec fix) Add a paragraph that requires `https://` servers returned by this discovery protocol to to ALWAYS  have  `Access-Control-Allow-Origin: *`  etc set up 
    - cc @guseggert for visibility, as we should include note about CORS headers in https://github.com/ipfs/specs/pull/337 too
1.  (more involved) Create libp2p version of https://github.com/ipfs/specs/pull/337 that browser peers could use over existing `/wss` or `/webtransport` listeners. Another argument [Why IPFS needs Delegated Routing over libp2p](https://www.notion.so/protocollabs/wip-Why-IPFS-needs-Delegated-Routing-over-libp2p-b634a4ec905c427fa41a97a3b3ba804e).

lidel commented on IPIP/0342-content-router-discovery.md:
```diff
@@ -0,0 +1,267 @@
+# IPIP 0342: Content Router Ambient Discovery
```
Since I can't expose a partial index without being punished, would this be more correct?

```suggestion
# IPIP 0342: IPNI Content Router Ambient Discovery
```

willscott commented on IPIP/0342-content-router-discovery.md:
```diff
@@ -0,0 +1,267 @@
+Pros:
+* Network is already there, no need to create a new protocol to "provide" new providers instead of CIDs.
+* You could potentially associate a provider with a specific root CID content.
+Cons:
+* Nodes cannot drop use of the DHT / other content routing options always are 'second tier'.
```
the alternative of other content routers being found in the DHT, which is what this alternative is trying to describe, does mean that no ipfs node could be run without kad dht code for DHT lookups. Being a DHT participant is more complexity than just having libp2p code to be able to connect to other peers, and at least my understanding of what's being proposed in this alternative is that it is intertwined with the DHT and not equivalent to hardcoded bootstrap nodes through which content routers can be learned.

willscott commented on IPIP/0342-content-router-discovery.md:
```diff
@@ -0,0 +1,267 @@
+new libp2p protocol advertised as "/ipfs/content-router-discovery/1.0.0".
+
+The protocol will follow a request-response model.
+A node will open a stream on the protocol when it wants to discover new
+content routers it does not already know.
```
updated in the section below around how a node should rank which other peers to query.

lidel commented on IPIP/0342-content-router-discovery.md:
```diff
@@ -32,6 +32,20 @@ is also insufficient long term because:
 
+This spec is designed for the ability of IPFS nodes to automatically discover
+and make use of 'content routers'. Content routers are services which are able
+to fulfill libp2p's [ContentRouting](https://github.com/libp2p/go-libp2p/blob/master/core/routing/routing.go#L26)
```
nit: this is not libp2p API, it is there due to legacy reasons (libp2p being part of go-ipfs long time ago).
afaik libp2p team removed specs for this from their repo and we have a gap  in IPFS one (see https://github.com/ipfs/specs/issues/345).

```suggestion
to fulfill IPFS [ContentRouting](https://github.com/libp2p/go-libp2p/blob/master/core/routing/routing.go#L26)
```

lidel commented on IPIP/0342-content-router-discovery.md:
```diff
@@ -32,6 +32,20 @@ is also insufficient long term because:
+to fulfill libp2p's [ContentRouting](https://github.com/libp2p/go-libp2p/blob/master/core/routing/routing.go#L26)
+API. These routers currently are considered to directly support queries using
+the protocols specified by
+[IPIP-337](https://github.com/ipfs/specs/pulls)
```
```suggestion
[IPIP-337](https://github.com/ipfs/specs/pull/337)
```

lidel commented on IPIP/0342-content-router-discovery.md:
```diff
@@ -32,6 +32,20 @@ is also insufficient long term because:
+This spec is designed for the ability of IPFS nodes to automatically discover
+and make use of 'content routers'. Content routers are services which are able
+to fulfill libp2p's [ContentRouting](https://github.com/libp2p/go-libp2p/blob/master/core/routing/routing.go#L26)
+API. These routers currently are considered to directly support queries using
+the protocols specified by
+[IPIP-337](https://github.com/ipfs/specs/pulls)
+and/or
+[IPIP-327](https://github.com/ipfs/specs/pull/327).
```
We need to either decide which one is the future, or document how client decides which one to use for sending requests.
IF we do the latter, including content type in response is the way to go: Reframe endpoint is  `application/vnd.ipfs.rpc[..]; version=n`.

willscott commented:
I updated to hopefully address your review, @lidel 
* There's a more concrete description of what the proposed protocol will look like
* I added a query tag to support discovery of more than just content routing through the mechanism
* in terms of 'is this only for ipni' / are you 'punished' if you're incomplete - i think it's probably too early to try to predict the dynamics here without either pretty extensive modeling or subsequent user data. The ranking / propagation of routers would be impacting by what users are after - so if there's a community or application that's focused on a narrow subset that's well addressed by a different router, that router could also be very viable with this discovery mechanism.

guillaumemichel commented on IPIP/0342-content-router-discovery.md:
```diff
@@ -0,0 +1,340 @@
+1. When it needs to perform a content routing query, and has not
+successfully performed a sync in over a day.
+2. When it's auto-nat status indicates it is eligible to be a DHT server, and
+it has not successfully performed a synce in over a day.
```
```suggestion
it has not successfully performed a sync in over a day.
```

guillaumemichel commented on IPIP/0342-content-router-discovery.md:
```diff
@@ -0,0 +1,340 @@
+nodes do not have geographic locality. As a result, performance is
+separated in the tracking of content routers because it will not be
+effective as a ranking factor in the non-geographically-aware
+gossip system described here. As an optimization, nodes may choose to
```
gossiping should be geographically-aware and happen only between peers that are geographically close to each other. Otherwise, I may share a content router that is geographically close from me, but it will be too slow for you, and you won't use it at all.
So sharing content routers with geographically far peers becomes irrelevant, as long as we had enough content routers and that they are distributed around the globe.

willscott replied:
do we believe ipfs nodes will generally have enough knowledge to ambiently identify which peers are geographically close to them?

Winterhuman replied:
https://github.com/libp2p/specs/pull/413 (GossipSub v1.2) would probably solve this, since it's all about minimising latency

willscott replied:
we could use the same modeling / structure as gossip sub, but this is meant to be pull-based rather than push based. I have concerns about dropping in GossipSub directly.

lidel replied:
Just noting that routers identified with DNS names  (cid.contact) could use things like Anycast to ensure the client is routed to the closest one. (I believe we already do that for ipfs.io gateway..)

guillaumemichel commented on IPIP/0342-content-router-discovery.md:
```diff
@@ -0,0 +1,340 @@
+
+Pros:
+* Network is already there, no need to create a new protocol to "provide" new providers instead of CIDs.
+* You could potentially associate a provider with a specific root CID content.
+Cons:
```
```suggestion
* You could potentially associate a provider with a specific root CID content.

Cons:
```

guillaumemichel commented on IPIP/0342-content-router-discovery.md:
```diff
@@ -0,0 +1,340 @@
+
+* a new provider is only visible to directly connected peers. they only forward it to peers asking them if it meets their bar
+for reliability.  This means propogation through the network is only posisble for routers that behave correctly.
+* because clients only propagate their 'top' routers, latency is also relevant, and with sufficient number of routers, the would only
```
```suggestion
* because clients only propagate their 'top' routers, latency is also relevant, and with sufficient number of routers, they would only
```

mxinden commented on IPIP/0342-content-router-discovery.md:
````diff
@@ -99,12 +107,50 @@ to content discovery queries. By default this will be 10. (This default is
+
+#### protocol messages
+
+Protocol messages are encoded using *cbor*. The following protocol examples demonstrate
+the schemas of requests and responses if they were to be encoded with JSON.
+
+A query on the "/ipfs/router-discovery/1.0.0" protocol will look like:
+```json
+{
+  "router": "string",
+  "filter": "bytes of the bloom filter"
+}
+```
````
> (we could have a separate type for discovering routers that support peer routing).

Thus they would serve the same use-case as a [rendezvous](https://github.com/libp2p/specs/commits/master/rendezvous/README.md) server?

> thoughts on the use case and the wire format here?

I can not think of a project outside of the IPFS realm that is in need for this.

willscott replied:
> Thus they would serve the same use-case as a [rendezvous](https://github.com/libp2p/specs/commits/master/rendezvous/README.md) server?

It extends the rendezvous protocol in two ways:
1. not requiring the need for a single hard-coded rendezvous point
2. adding reputational gossip in addition to just directory listing in the rendezvous protocol

guseggert commented on IPIP/0342-content-router-discovery.md:
```diff
@@ -0,0 +1,341 @@
+and/or
+[IPIP-327](https://github.com/ipfs/specs/pull/327).
+
+In addition, this protocol expects that content routers that may be considered
+for auto-configuration/discovery by IPFS nodes will have knowledge of the
+entire CID space - in other words a delegation to such a router may be
+considered 'exhaustive'.
```
How does this happen? What kind of consistency SLAs should indexers have, and how can they achieve it?

willscott replied:
I'd like to say 'that's outside of this direct IPIP' - in that if routers fail to be consistent they would risk loosing priority.

In practice:
* indexers follow the list of providers from other indexers so that the constituents they follow are consistent
* they gossip announcements they see to each other so new updates are propagated between them
* [in progress] they can come to snapshot consensus periodically over a vector of providers & latest advertisements.

BigLep commented:
This came up in https://pl-strflt.notion.site/2023-05-30-Content-Routing-WG-12-b2ed74834fe44e359bbcdd02740e2084

There is going to be implementation need for this in the next quarter or two.  As a result, want to get ahead before code gets written and decisions ossify.

Some next steps:
1. Update the IPIP to [the newest template](https://github.com/ipfs/specs/blob/main/ipip-template.md) standard with lessons learned the last 6 months
  - I assume @willscott or someone from PL EngRes Bedrock team will do this.
2. Review from spec maintainers
  - I assume @lidel will take the first pass here on the updated doc when it's ready.

Implementation notes:
"Lassie currently does not depend on boxo. It may be hard for us to prototype ambient-discovery in boxo until the boxo/rest-of-world dependency conflicts are resolved"

lidel commented on IPIP/0342-content-router-discovery.md:
```diff
@@ -0,0 +1,341 @@
+  has served useful content in the past.
+* Latency / ping time of the peer.
+
+### 3. selection of routers
```
Note to self: this section of the spec should be more specific about "bare minimum reputation system", and provide enough for implementer to do the right thing, and not say clients do "as they wish".

Expected behavior on non-client services like bootstrappers should be also elaborated.

lidel commented on IPIP/0342-content-router-discovery.md:
```diff
@@ -0,0 +1,341 @@
+
+## Test fixtures
+
+TK is a CID currently only available through the content routing system,
+and not through the IPFS DHT. This is a piece of content that can be queried
+to validate the presence of alternative content routing systems.
```
Can't hardcode CID in the spec, it will stop being useful the moment someone announces it on DHT.

```suggestion
An unique CID needs to be created every time and made only available through the tested router, 
and not through any of the other ones, including the IPFS DHT. 

This piece of content can be queried to validate the presence of alternative content routing systems.
```

willscott replied:
pointed to a implementation of a provider that can generate unique cid and provide it only over a configured IPNI endpoint as a basis for testing.

lidel commented on IPIP/0342-content-router-discovery.md:
````diff
@@ -0,0 +1,357 @@
+}
+```
+
+A response is a list of entries, which looks like:
+
+```json
+[
+  {
+    "peer": "multiaddr.MultiAddr",
+    "score": float
````
- single `score` may be too vague. based on Rhea/Saturn alone,  we may want to track "lookup" score and "retrieval" score separately
- if we want this API to be not limited to IPNI, then we could have `type` per result

SgtPooki commented on src/ipips/ipip-0342.md:
```diff
@@ -0,0 +1,366 @@
+to fulfill IPFS's [ContentRouting](https://github.com/libp2p/go-libp2p/blob/master/core/routing/routing.go#L26)
+API. These routers currently are considered to directly support queries using
+the protocols specified by
+[IPIP-337](https://github.com/ipfs/specs/pulls/337)
+and/or
+[IPIP-327](https://github.com/ipfs/specs/pull/327).
```
Would it be worth calling out these names as well, so users unfamiliar with the ipip-### dont have to click through

SgtPooki commented on src/ipips/ipip-0342.md:
```diff
@@ -0,0 +1,366 @@
+and/or
+[IPIP-327](https://github.com/ipfs/specs/pull/327).
+
+In addition, this protocol expects that content routers that may be considered
+for auto-configuration/discovery by IPFS nodes will have knowledge of the
+entire CID space - in other words a delegation to such a router may be
+considered 'exhaustive'.
+
```
Is it possible to have an exhaustive list of the entire CID space?

Also, the wording here is hard to follow. Maybe something like:

```suggestion
In addition, implementers of this protocol may specify default content routers, but this protocol comes with a few expectations for default content-routers:

Default content routers configured by those implementing this protocol:

* MUST know of the
entire CID space - in other words, a delegation to such a router may be
considered 'exhaustive'

```

guillaumemichel replied:
I also doubt that any single/federated entity will know the entire CID space. It may be possible now, but this certainly doesn't scale.

If we shard the CIDs on different instances, we basically get a DHT (with O(1) lookup if all shards fit in memory). However, it inherits the same weaknesses that the current DHT has, when providing a fair amount of CIDs, you need to open a connection to every single shard. To mitigate this, we might consider introducing intermediary services tasked with CID-to-shard allocation, streamlining access and reducing the number of direct connections needed.

SgtPooki commented on src/ipips/ipip-0342.md:
```diff
@@ -0,0 +1,366 @@
+
+### 0. content-router discovery state tracking
+
+Nodes will conceptually track a registry about known content routers.
+This registry will be able to understand for a given content router two
+properties:
```
```suggestion
Nodes will conceptually track a registry of known content routers.
This registry will maintain two properties for a given content router:
```

SgtPooki commented on src/ipips/ipip-0342.md:
```diff
@@ -0,0 +1,366 @@
+Nodes will conceptually track a registry about known content routers.
+This registry will be able to understand for a given content router two
+properties:
+* reliability - how many good vs bad responses has this router responded
+with. This statistic should be windowed, such that the client can calculate
+it in terms of the last week or month. This will in practice be stored as
+daily buckets of successful and unsuccessful queries against a router, where
+success indicates that the router was queried, and the data was subsequently
+retrieved from a node returned as a provider by that router.
```
Is it possible for a content router who returns a peer that *should* have content, to know that they successfully provided that content?

1. A asks B for providers of bafy1
2. B finds C and tells A about it
3. A asks C for bafy1.

Is this spec requiring A to tell B the result of item 3 above? How is this tracking accomplished?

guillaumemichel commented on src/ipips/ipip-0342.md:
```diff
@@ -0,0 +1,366 @@
+that it has a greater than 99% certainly that it will receive a useful
+response. The maximum size of a query may be capped by the server, but can be
+effectively considered to be under 10kb.
+* The client will hash it's known content routers into the bloom filter
```
typo

```suggestion
* The client will hash its known content routers into the bloom filter
```

guillaumemichel commented on src/ipips/ipip-0342.md:
```diff
@@ -0,0 +1,366 @@
+* The server will have a parameter for a number of servers it wants to return
+to content discovery queries. By default this will be 10. (This default is
+picked as the result of modeling router propagation). It will iterate through
+it's list of known content routers, hashing them against the bloom filter and
```
```suggestion
its list of known content routers, hashing them against the bloom filter and
```

guillaumemichel commented on src/ipips/ipip-0342.md:
```diff
@@ -0,0 +1,366 @@
+picked as the result of modeling router propagation). It will iterate through
+it's list of known content routers, hashing them against the bloom filter and
+selecting the top routers that are not already known to the client. It will
+return this list, along with it's reliability score for each.
```
```suggestion
return this list, along with its reliability score for each.
```

guillaumemichel commented on src/ipips/ipip-0342.md:
```diff
@@ -0,0 +1,366 @@
+
+### 2. probing of the discovery protocol
+
+A node will probe it's connected peers for content routing updates in two
```
```suggestion
A node will probe its connected peers for content routing updates in two
```

guillaumemichel commented on src/ipips/ipip-0342.md:
```diff
@@ -0,0 +1,366 @@
+
+1. When it needs to perform a content routing query, and has not
+successfully performed a sync in over a day.
+2. When it's AutoNAT status indicates it is eligible to be a DHT server, and
```
```suggestion
2. When its AutoNAT status indicates it is eligible to be a DHT server, and
```

guillaumemichel commented on src/ipips/ipip-0342.md:
```diff
@@ -0,0 +1,366 @@
+of local selection.
+They are maintained for a month for the purpose of admitting
+knowledge of routers to others - so a client will no longer set bits for
+routers it is aware of but which do not meet it's threshold for 'good'
```
```suggestion
routers it is aware of but which do not meet its threshold for 'good'
```

guillaumemichel commented on src/ipips/ipip-0342.md:
```diff
@@ -0,0 +1,366 @@
+
+1. When it needs to perform a content routing query, and has not
+successfully performed a sync in over a day.
+2. When its AutoNAT status indicates it is eligible to be a DHT server, and
+it has not successfully performed a sync in over a day.
```
What is the reasoning here?
